### PR TITLE
Process data

### PR DESCRIPTION
**Pull request by Said-laasri**
**Create models**

- Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
- Set up associations between models.

**Use models to insert data**

- Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
- Open rails console for your project.
- Create at least one user by running the following code:
-   ` first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')`
-    `second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')`
- Create at least 4 posts written by one of the users you created by running the following code:
-   ` first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')`
- Create at least 6 posts comments for one of the posts you created by running the following code:
-    `Comment.create(post: first_post, author: second_user, text: 'Hi Tom!' )`
- Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

**Create custom methods**
Your models should include some additional methods.

1. Users

A method that returns the 3 most recent posts for a given user.

2. Posts

A method that updates the posts counter for a user.
A method which returns the 5 most recent comments for a given post.

3. Comments

A method that updates the comments counter for a post.

4. Likes

A method that updates the likes counter for a post.
Go to rails c and check if your methods are working.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts
+  has_many :likes
+  has_many :comments
```
Comment (rloterh) [inline app/models/user.rb]: - [x] As per the given _ERD_, the _User_ **association** should have _author_id_ referenced as **foreign key** in _Post_, _Comment,_ and _Like_ models. Kindly ensure to improve this aspect, as can be seen below:
```ruby
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
```
Comment (rloterh) [inline app/models/comment.rb]: - [ ] As per the given _ERD_, the _Comment_ **association** should have _Comment_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :user
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] As per the given _ERD_, the _Post_ **association** should have _Post_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
```
Comment (rloterh) [inline app/models/like.rb]: - [x] As per the given _ERD_, the _Like_ **association** should have _Like_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/like.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+
+  def update_counter
```
Comment (rloterh) [inline app/models/like.rb]: - [x] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the update_likes_counter method every time a new like has been saved to database, then make the `update_likes_counter` method private, similar to:

```ruby
 # app/models/likes.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

File: app/models/post.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  has_many :comments
+  has_many :likes
+
+  def update_posts_counter
```
Comment (rloterh) [inline app/models/post.rb]: - [x] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_posts_counter method every time a new like has been saved to database, then make the `update_likes_counter` method private, similar to:

```ruby
 # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+  def update_counter
```
Comment (rloterh) [inline app/models/comment.rb]: - [x] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_comments_counter method every time a new like has been saved to database, then make the `update_likes_counter` method private, similar to:

```ruby
 # app/models/comment.rb
  after_save :update_comment_counter

  private

  def update_comment_counter
    author.increment!(:comments_counter)
  end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Said, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- No linter errors

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Kindly check the comments marked **[OPTIONAL]** under the review._

**You can also consider:**
- I think you should change the header title _Authors_ to _Author_ in your ReadMe, as this project is worked on by a sole contributor. 
Implementing the above change(s) will make your ReadMe/ repository stand out even better 😉

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 1 above)
Comment (Said-laasri) [reply app/models/user.rb]: @rloterh  Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/like.rb (hunk 5 above)
Comment (Said-laasri) [reply app/models/like.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/post.rb (hunk 6 above)
Comment (Said-laasri) [reply app/models/post.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/comment.rb (hunk 7 above)
Comment (Said-laasri) [reply app/models/comment.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/like.rb (hunk 4 above)
Comment (Said-laasri) [reply app/models/like.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/post.rb (hunk 3 above)
Comment (Said-laasri) [reply app/models/post.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

File: app/models/comment.rb (hunk 2 above)
Comment (Said-laasri) [reply app/models/comment.rb]: @rloterh Thank you for reviewing my project, I will make the necessary corrections as suggested

**Review by ShahierNashaat — APPROVED** [review]
Hi @Said-laasri, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 💯 

- Correct use of GitFlow ✔️
- Descriptive commit messages ✔️
- Professional looking pull request ✔️
- Your readme looks professional ✔️
- Your code is clean and well documented ✔️

# Approved :ok_hand: :ok_hand: :ok_hand:

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Said-laasri** [comment]
@ShahierNashaat Thank you for reviewing my project

**Pull request merged**